src/seraphis_mocks/scan_context_async_mock.cpp

```diff
@@ -697,36 +725,51 @@ void AsyncScanContextLegacy::wait_until_pending_queue_clears()
 
     m_pending_chunk_queue.shut_down();
 
-    PendingChunk clear_chunk;
-    async::TokenQueueResult clear_chunk_result = m_pending_chunk_queue.force_pop(clear_chunk);
-    while (clear_chunk_result != async::TokenQueueResult::SHUTTING_DOWN)
+    // Send stop signals to all pending tasks and drain the queue
+    std::vector<PendingChunk> drained_chunks;
+    PendingChunk pending_chunk;
+    auto pending_chunk_res = m_pending_chunk_queue.force_pop(pending_chunk);
+    while (pending_chunk_res != async::TokenQueueResult::SHUTTING_DOWN)
     {
-        THROW_WALLET_EXCEPTION_IF(clear_chunk_result != async::TokenQueueResult::SUCCESS,
-            tools::error::wallet_internal_error, "Failed to clear onchain chunks");
+        THROW_WALLET_EXCEPTION_IF(pending_chunk_res != async::TokenQueueResult::SUCCESS,
+            tools::error::wallet_internal_error, "Failed to clear pending chunks");
 
         // Send stop signals
-        clear_chunk.pending_context.stop_signal.set_value();
-        clear_chunk.pending_data.stop_signal.set_value();
+        pending_chunk.pending_context.stop_signal.set_value();
+        pending_chunk.pending_data.stop_signal.set_value();
+
+        // Push the pending chunk into our drain catcher
+        drained_chunks.push_back(std::move(pending_chunk));
+
+        // Get the next pending chunk if there is one
+        pending_chunk_res = m_pending_chunk_queue.force_pop(pending_chunk);
+    }
+
+    // Wait until all work from the queue is done
+    while (!drained_chunks.empty())
+    {
+        auto &clear_chunk = drained_chunks.back();
+        MDEBUG("Waiting to clear onchain chunk starting at " << clear_chunk.chunk_request.start_index);
 
-        // Wait until all work in the pending queue is done, not just contexts
-        // TODO: wait until every task in the pool has returned
+        // Wait until **data** join condition is set, we're not waiting on just the contexts
         m_threadpool.work_while_waiting(clear_chunk.pending_data.data_join_condition,
             async::DefaultPriorityLevels::MAX);
 
-        clear_chunk_result = m_pending_chunk_queue.force_pop(clear_chunk);
+        drained_chunks.pop_back();
     }
```

Comment:
The only way it can return  `async::TokenQueueResult::SHUTTING_DOWN` is if the queue is empty

https://github.com/UkoeHB/monero/blob/c7422607385a070d9e0069f96598be78336f1ef8/src/async/token_queue.h#L109-L110

Since this function is holding `m_async_scan_context_mutex`, and we've now made sure to work until all tasks are done, nothing can theoretically exist that uses `force_push` to insert a task in parallel at this point in this function

Reply:
Writing that comment^ made me re-check the "gap fill" edge case which wasn't checking `m_scanner_ready` before `force_push`. Done. That comment should hold now^